### PR TITLE
Validator integration with current metrics processor for logging

### PR DESCRIPTION
**Pull request by wesleytruong**
Integrated the validator together with metrics processor for better metrics logging.

Key changes:
- Metrics processor is passed to validator within training loop
- Validator can reuse metrics processor's built-in functionalities such as memory profiling, throughput tracking, and tensorboard/wandb logging

This is how the new logging looks from terminal:
Screenshot 2025-07-14 at 3 22 56 PM

### CODE REVIEW

File: torchtitan/components/validate.py [hunk 1]
```diff
@@ -53,6 +53,7 @@ def __init__(
         loss_fn: LossFunction,
         validation_context: Generator[None, None, None],
         maybe_enable_amp: Generator[None, None, None],
+        metrics_processor: MetricsProcessor | None = None,
```
Comment (tianyu-l) [inline torchtitan/components/validate.py]: It seems you assume it's not `None` anyway. So let's just do
```suggestion
        metrics_processor: MetricsProcessor,
```

File: torchtitan/components/metrics.py [hunk 2]
```diff
@@ -415,6 +417,39 @@ def log(
         self.time_last_log = time.perf_counter()
         self.device_memory_monitor.reset_peak_stats()
 
+    def log_validation(self, val_loss: float, step: int):
```
Comment (tianyu-l) [inline torchtitan/components/metrics.py]: nit comment: maybe enough to just
```suggestion
    def log_validation(self, loss: float, step: int):
```

File: torchtitan/components/metrics.py [hunk 3]
```diff
@@ -415,6 +417,39 @@ def log(
...
+        time_delta = time.perf_counter() - self.time_last_log
+
+        device_mem_stats = self.device_memory_monitor.get_peak_stats()
+
+        # tokens per second per device, abbreviated as tps
+        tps = self.validation_ntokens_since_last_log / (
+            time_delta * self.parallel_dims.non_data_parallel_size
+        )
+
+        val_metrics = {
```
Comment (tianyu-l) [inline torchtitan/components/metrics.py]: nit: to be consistent
```suggestion
        metrics = {
```

File: torchtitan/components/metrics.py [hunk 4]
```diff
@@ -307,6 +307,7 @@ class MetricsProcessor:
 
     gpu_peak_flops: int
     ntokens_since_last_log: int
+    validation_ntokens_since_last_log: int
```
Comment (tianyu-l) [inline torchtitan/components/metrics.py]: since we are reusing other fields anyway, I think we can just reuse `ntokens_since_last_log`

File: torchtitan/components/validate.py [hunk 5]
```diff
@@ -126,9 +130,7 @@ def validate(
         else:
             global_avg_loss = loss
```
Comment (tianyu-l) [inline torchtitan/components/validate.py]: I think here you need the following to send loss from GPU back to CPU before you can even print them out. This is the branch when no FSDP/CP is used (i.e. during TP / PP / single-device training)
```suggestion
            global_avg_loss = loss.item()
```

**Review by tianyu-l — COMMENTED** [review]
Looks great, had some comments.

**Comment by wesleytruong** [comment]
@tianyu-l addressed the comments

**Review by tianyu-l — APPROVED** [review]
looks very neat, great job!
